src/optimization/inlineConstructors.ml

```diff
@@ -42,47 +41,69 @@ open Globals
 	| IOKArray of int
 
+(*
+	inline_object
+	Represents an instance of TNew TObjectDecl or TArrayDecl that has potential to be inlined.
+	Wether the inlining is cancelled or not is decided during the analysis phase.
+*)
 and inline_object = {
 	io_kind : inline_object_kind;
-	io_expr : texpr;
-	io_pos : pos;
-	io_has_untyped : bool;
-	mutable io_cancelled : bool;
-	mutable io_declared : bool;
-	mutable io_aliases : inline_var list;
-	mutable io_fields : (string,inline_var) PMap.t;
-	mutable io_id_start : int;
-	mutable io_id_end : int;
+	io_expr : texpr; (* This is the inlined constructor expression *)
+	io_pos : pos; (* The original position of the constructor expression *)
+	mutable io_has_untyped : bool; (* Wether inlining this object would bring untyped expressions into the parent expression *)
+	mutable io_cancelled : bool; (* Wether this inline object has been cancelled *)
+	mutable io_declared : bool; (* Wether the variable declarations for this inline object's fields have already been output. (Used in final_map) *)
+	mutable io_aliases : inline_var list; (* List of variables that are aliasing/referencing this inline object *)
+	mutable io_fields : (string,inline_var) PMap.t; (* The fields that this inline object supports, fields are inline variables which might alias other inline_objects *)
+	mutable io_inline_methods : texpr list; (* List of inlined method calls. Populated during analysis and consumed by the final_map phase *)
+	mutable io_dependent_vars : tvar list; (* List of variables that should be cancelled if this inline object is cancelled *)
 }
 
 and inline_var_kind =
-	| IVKField of inline_object * string * texpr option
+	| IVKField of
+		inline_object *
+		string * (* The field name *)
+		texpr option (* If this is Some _ then this field is a constant. (Used for Array .length property) *)
 	| IVKLocal
 
 and inline_var_state =
-	| IVSUnassigned
-	| IVSAliasing of inline_object
-	| IVSCancelled
+	| IVSUnassigned (* The variable isn't yet assigned *)
+	| IVSAliasing of inline_object (* The variable is aliasing an inline object *)
+	| IVSCancelled (* The variable was cancelled and should no longer be considered for aliasing inline objects *)
 
+(*
+	inline_var
+	Represents a local variable that is tracked by the inline constructor analysis.
+	It's main purpose is to track variables that are considered aliases of inline objects.
+	It also tracks all unassigned variables in the program.
+*)
 and inline_var = {
 	iv_var : tvar;
 	mutable iv_state : inline_var_state;
 	mutable iv_kind : inline_var_kind;
-	mutable iv_closed : bool
+	mutable iv_closed : bool (* Inline variables are marked as closed when the scope they were first assigned on ends, any appearance of this variable after it has been closed causes cancellation *)
 }
 
+and inline_object_field = 
+	| IOFInlineMethod of inline_object * inline_var * tclass * Type.tparams * tclass_field * tfunc
+	| IOFInlineVar of inline_var
+	| IOFNone
+
 let inline_constructors ctx e =
```

Comment:
Most methods in Haxe code probably don't contain inlinable objects/arrays. 
Wouldn't it be more efficient to run `check_expr` on `e` before doing any actual work to find out if the expression has something to inline.

Reply:
This is a good idea.

Instead of `check_expr`, wouldn't it be better to check if `curr_io_id` is not 0?
In contrast to `check_expr` in the case that no inline objects are found it will have the cost of a noop Type.map_expr. But when objects are found it saves the cost of an extra iteration through the expression tree.

Reply:
On second thought, using `check_expr` is probably better since it can short circuit and ends as soon as one inline object is found.

Reply:
Yes, that's what I'm talking about.

Reply:
I've committed this change.